applications/StructuralMechanicsApplication/python_scripts/structural_response.py

```diff
@@ -220,7 +220,215 @@ def GetShapeGradient(self):
+
+
+    def InitializeSolutionStep(self):
+        # synchronize the modelparts # TODO this should happen automatically
+        print("\n> Synchronize primal and adjoint modelpart for response:", self.identifier)
+        for node in self.adjoint_model_part.Nodes:
+            ref_node = self.primal_model_part.Nodes[node.Id]
+            node.X0 = ref_node.X0
+            node.Y0 = ref_node.Y0
+            node.Z0 = ref_node.Z0
+            node.X = ref_node.X
+            node.Y = ref_node.Y
+            node.Z = ref_node.Z
+
+
+        # Run the primal analysis.
+        # TODO if primal_analysis.status==solved: return
+        print("\n> Starting primal analysis for response:", self.identifier)
+        startTime = timer.time()
+        self.primal_analysis.time = self.primal_analysis._GetSolver().AdvanceInTime(self.primal_analysis.time)
+        self.primal_analysis.InitializeSolutionStep()
+        self.primal_analysis._GetSolver().Predict()
+        self.primal_analysis._GetSolver().SolveSolutionStep()
+        self.primal_analysis.FinalizeSolutionStep()
+        self.primal_analysis.OutputSolutionStep()
+        print("> Time needed for solving the primal analysis = ",round(timer.time() - startTime,2),"s")
+
+        # TODO the response value calculation for stresses currently only works on the adjoint modelpart
+        # this needs to be improved, also the response value should be calculated on the PRIMAL modelpart!!
+        self.adjoint_analysis.time = self.adjoint_analysis._GetSolver().AdvanceInTime(self.adjoint_analysis.time)
+        self.adjoint_analysis.InitializeSolutionStep()
+
+    def CalculateValue(self):
+        startTime = timer.time()
+        value = self.response_function_utility.CalculateValue(self.adjoint_model_part)
+        print("> Time needed for calculating the response value = ",round(timer.time() - startTime,2),"s")
+
+        self.primal_model_part.ProcessInfo[StructuralMechanicsApplication.RESPONSE_VALUE] = value
+
+
+    def CalculateGradient(self):
+        print("\n> Starting adjoint analysis for response:", self.identifier)
+        startTime = timer.time()
+        self.adjoint_analysis._GetSolver().Predict()
+        self.adjoint_analysis._GetSolver().SolveSolutionStep()
+        self.adjoint_analysis.FinalizeSolutionStep()
+        self.adjoint_analysis.OutputSolutionStep()
+        print("> Time needed for solving the adjoint analysis = ",round(timer.time() - startTime,2),"s")
+
+
+    def GetValue(self):
+        return self.primal_model_part.ProcessInfo[StructuralMechanicsApplication.RESPONSE_VALUE]
+
+
+    def GetShapeGradient(self):
+        gradient = {}
+        for node in self.adjoint_model_part.Nodes:
+            gradient[node.Id] = node.GetSolutionStepValue(SHAPE_SENSITIVITY)
+        return gradient
+
+
+    def FinalizeSolutionStep(self):
+        # TODO reset DISPLACEMENT, ROTATION ADJOINT_DISPLACEMENT and ADJOINT_ROTATION
```

Comment:
self.adjoint_analysis.FinalizeSolutionStep()

Reply:
'adjoint_analysis.FinalizeSolutionStep()' has to be part of the `CalculateGradient` function, because the final response gradient is calculated there. The same way e.g. 'primal_analysis.FinalizeSolutionStep()' is done already in the `InitializeSolutionStep` of the response. because it concludes the primal analysis which is a prerequisite for the CalculateValue or CalculateGradient.

Reply:
The `primal_analysis.FinalizeSolutionStep()` is ok because it is guaranteed to be called in the logical order, i.e.

```
        InitializeSolutionStep()
        Predict()
        SolveSolutionStep()
        FinalizeSolutionStep()
        OutputSolutionStep()
```

This is because the InitializeSolutionStep(), ..., FinalizeSolutionStep() are at the same level of abstraction, i.e., they are called within the same function. My problem with only calling one or the other method (e.g., FinalizeSolutionStep() in CalculateGradient()) is that you no longer have this guarantee. Looking back at your code, I don't see a big problem to leave FinalizeSolutionStep empty, but you should include `self.adjoint_analysis.InitializeSolutionStep()` in the `CalculateGradient` function.

Reply:
This is part of the requested changes of the current review.

Reply:
Done, more consistent now